guide/developing-ui.md

```diff
@@ -0,0 +1,146 @@
+of Javascript files, css files, images, etc. These are "packed" together
+using [webpack](https://webpack.js.org).
+
+For ongoing development it is possible to have the UI running in a web
+browser using "hot loading". This means that as you make changes to source
+files, they will be detected, the webpack will be recompiled and generated
+and then the browser will be informed to reload things. In combination with
+the debugger built into modern browsers this makes for a reasonable process.
+
+For a production build, the same process is followed, except with different
+settings. In particular, no hot loading development server will be started
+and more effort is expended on packing and minimizing the components of
+the application as represented in the various "bundles". Read more about
+this in the webpack documentation.
+
+## Getting started
+
+Checkout the branch you want to work on and type
+
+    $ yarn start
+
+This will pack and prepare a development setup. By default the development
+server that serves up the web page(s) will listen on
+[http://localhost:3000](http://localhost:3000) so you can direct your browser
+there.
+
+Make any changes to the source code as you desire (look at existing code
+for examples and typical style), and the browser will hot-load your changes.
+Often times you will make mistakes. Anything from a coding error (for which
+you can use the browser's debugger), or compilation breaking Javascript errors.
+The latter will often be reported as errors during the webpack assembly
+process, but some will show up in the browser console, or both.
+
+## Control and location of settings
+
+Much of the settings needed to put the UI together, run webpack etc. is
+located in a series of files that contain webpack configuration. These
+files all live in the "webpack" subdirectory. We will not explain all
+of them here, as you should rarely need to understand them, let alone
+modify them.
+
+What is worth mentioning is that the `package.json` file is configured
+to use a different webpack configuration for development vs. production
+builds. Both configurations depend on a shared configuration common to
+both.
+
+There are also some settings that control aspects of the MoonFire UI
+behavior, such as window titles etc. These settings are found in the
+`settings-nvr.js` file in the project root directory. They should be
+pretty self explanatory.
+
+The webpack configuration for all webpack builds is able to load the
+values from `settings-nvr.js` and then, if the file exists, load those
+from `settings-nvr-local.js` and use them to add to the configuration,
+or replace. You can take advantage of this to add your own configuration
+values in the "local" file, which does not get checked in, but is used
+to affect development, and production builds.
+
+## Special considerations for API calls
+
+The UI code will make calls to the MoonFire NVR's server API, which is
```

Comment:
I'll do the substitution in a moment.

As far as the other questions: Yes, surprised too that there is no such thing. The approach sourcing additional/overriding env variables can work, in a basic manner. For settings I prefer keeping production and development together so you can see the differences and only have to know about one file.

It also does not allow the concept of mainline definitions and conditional overrides based on the web pack --mode parameter, at least not easily.

The approach I followed also allows all that, and allows controlling individual elements of settings, such as separating out the host and the port. Makes it less likely to have misconfigurations that are hard to diagnose.

For example, my local settings contains:
```
  webpack_mode: {
    development: {
      moonfire: {
        server: '192.168.10.232',
      },
    },
  },
```
Just to change my API proxy. I do not have to bother duplicating a port setting. Thus, if the default changes, this will still work. With a complete URL, changing the default port elsewhere, would not change a port included in this URL and problems...